### ISSUE TITLE
CPU training stalls for 60s+ compiling native artifact for KNN predictor with batch normalization and quantized activation on backend cpu

### ISSUE THREAD

**Issue by nm-z** [nm-z/recipe-dev#303]
Public `recipe.model().knn(3).tanh().norm(batch).qi(3).k.s.knn(3).leak().norm(batch).iq(3).s.loss(mae)` stalls at 60 s on `backend=cpu`. Earliest failing operation is `.knn(3).tanh().norm(batch).qi(3).k.s`: it lowers through `fit_knn` to `Predictor` (`Primitive::Predictor`) combined with `.tanh()` (`Activation::Tanh`), `.norm(batch)` (`Primitive::Normalize`, batch mode), and `.qi(3).k.s` integer quantization encoding, then stalls at `compile_native_artifact` / `native_cpu_compiler` (recipe.rs:2824-2831). The temporal shard dataset (~30 shards, ~96 rows each, 1 feature) and 3-NN estimator arithmetic justify a sub-10 s epoch; measured 60 s stable exceeds workload and hardware justification. No existing open issue covers the KNN predictor kernel with batch normalization and quantized activation on CPU (254 is Bayes/perceptron, 255 Forest/XGBoost, 287 CatBoost fit, 278 Pool, 290 residual layer). The second `.knn(3)...` stage and `loss(mae)` are downstream and do not change the earliest cause.

## Machine provenance

- Discoverer: Recipe exhaustive Cartesian traversal
- Failure fingerprint: `f260da5b5850e954`
- Classifier provider: `openrouter`
- Classifier model: `thinkingmachines/inkling-small:free`
- Classifier effort: `default`

### Classifier decision

```json
{"verdict":"new","issue":0,"title":"CPU training stalls for 60s+ compiling native artifact for KNN predictor with batch normalization and quantized activation on backend cpu","body":"Public `recipe.model().knn(3).tanh().norm(batch).qi(3).k.s.knn(3).leak().norm(batch).iq(3).s.loss(mae)` stalls at 60 s on `backend=cpu`. Earliest failing operation is `.knn(3).tanh().norm(batch).qi(3).k.s`: it lowers through `fit_knn` to `Predictor` (`Primitive::Predictor`) combined with `.tanh()` (`Activation::Tanh`), `.norm(batch)` (`Primitive::Normalize`, batch mode), and `.qi(3).k.s` integer quantization encoding, then stalls at `compile_native_artifact` / `native_cpu_compiler` (recipe.rs:2824-2831). The temporal shard dataset (~30 shards, ~96 rows each, 1 feature) and 3-NN estimator arithmetic justify a sub-10 s epoch; measured 60 s stable exceeds workload and hardware justification. No existing open issue covers the KNN predictor kernel with batch normalization and quantized activation on CPU (254 is Bayes/perceptron, 255 Forest/XGBoost, 287 CatBoost fit, 278 Pool, 290 residual layer). The second `.knn(3)...` stage and `loss(mae)` are downstream and do not change the earliest cause.","rationale":"Searched open performance issues (254, 255, 250, 278, 290, 287, 160, 193, 282). The reproduction reaches `.knn(3).tanh().norm(batch).qi(3).k.s` through `fit_knn` -> `Predictor` with batch normalization and `.qi(3).k.s` quantization, stalling in CPU native artifact compilation. That earliest operation and implementation path (`lower_estimator` / `compile_native_artifact`) are not represented by any existing open issue; closest issues use different predictor primitives (Bayes/perceptron, Forest/XGBoost, CatBoost, Pool, residual, kmeans estimation). The dataset is small (~3k rows, 1 feature) so 60 s is materially slower than justified throughput. Thus new loader/compilation defect, not a comment."}
```

## Deterministic failure packet

kind=performance
id=f260da5b5850e954
base=commit=5306c211d9fc2b0186a4eaf6d9e25a10caec49ca tracked_tree=modified
cursor=seed:0 cursor:10844 next:10845 composition:112909307072596
expected=one training epoch uses throughput proportionate to its operations, data volume, memory traffic, and available hardware
observed=the public Recipe run remained active after 60 seconds
measurement=elapsed_seconds:60 backend:cpu
replay=the configured slow-runtime threshold was crossed stable:true
command=cargo run --bin recipe -- /home/nate/Desktop/recipe-issue-machine/target/debug/recipe-trials/recipe-trial-181-10844/reproduction.rs
reproduction:
```rust
use recipe::*;

fn main() {
	let bundle = format!("/tmp/recipe-composition-repro-{}.ogdl", std::process::id());
	let data = recipe.data("data/temporal/daily_shards")
		.target("target")
		.split(0.8);
	let model = recipe.model()
		.knn(3).tanh().norm(batch).qi(3).k.s
		.knn(3).leak().norm(batch).iq(3).s
		.loss(mae);
	let report = recipe.train()
		.optimizer(adamw)
		.lr(0.0001)
		.seed(112909307072596)
		.epochs(1)
		.log(all)
		.stop(0.8)
		.fp(16)
		.save(&bundle)
		.run(&model, &data);
	assert!(report.final_loss().is_finite());
}
```
backend=cpu

**Comment by nm-z** [comment]
Closing as resolved based on zero-change verification PR #422.

That PR changes 0 files and verifies the existing `minimal` behavior, so there is no code to merge. The original verification is copied below.

<details>
<summary>Verification from PR #422: Verify current CPU quantized KNN compilation</summary>

Closes #303.

The exact public issue source now completes through both CPU native artifact states:

- Cold path: 2.37 seconds. Five CPU artifacts were compiled from cache misses, the real training epoch took 193.666 ms, and training MAE was finite at 20.7500.
- Warm path: 1.00 second. Every CPU artifact loaded from cache, the real training epoch took 192.060 ms, and training MAE was again 20.7500.
- Both runs trained 2,304 rows from data/temporal/daily_shards and completed holdout evaluation. The saved disk-backed bundle is 413,935 bytes.

The current Clang CPU artifact path completes far below the reported 60-second threshold. The previous attempt for this report added diagnostic timing only and was never merged.

This branch intentionally has one verification commit and no tree diff because current evidence does not support another production-code change.

</details>

Accepted verdict: PASS

Reason: The verification above reports that current `minimal` completes this issue's reported path successfully.
